WireDomain/Sources/WireDomain/Event Decryption/MLSMessageDecryptor.swift

```diff
@@ -0,0 +1,168 @@
+            WireLogger.mls.warn(
+                "failed to add mls message: conversation is not ready"
+            )
+
+            throw MLSMessageDecryptorError.mlsConversationNotReady
+        }
+
+        let decryptionResults = await decryptMLSMessage(
+            message: eventData.message,
+            mlsGroupID: mlsGroupID,
+            subconversation: eventData.subconversation
+        )
+
+        let decryptedMessages = await processMLSMessageDecryptionResults(
+            decryptionResults,
+            mlsConversation: mlsConversation,
+            senderID: eventData.senderID.uuid,
+            senderDomain: eventData.senderID.domain,
+            date: eventData.timestamp
+        )
+
+        var decryptedEvent = eventData
+        decryptedEvent.decryptedMessages = decryptedMessages
+
+        return decryptedEvent
+    }
+
+    private func decryptMLSMessage(
+        message: String,
+        mlsGroupID: MLSGroupID,
+        subconversation: String?
+    ) async -> [MLSDecryptResult] {
+        do {
+            let subconvType = subconversation != nil ? SubgroupType(rawValue: subconversation!) : nil
+
+            let results = try await mlsDecryptionService.decrypt(
+                message: message,
+                for: mlsGroupID,
+                subconversationType: subconvType
+            )
+
+            if results.isEmpty {
+                WireLogger.mls.info(
+                    "successfully decrypted mls message but no result was returned"
+                )
+
+                return []
+            }
+
+            return results
+
+        } catch {
+            WireLogger.mls.warn(
+                "failed to decrypt mls message: \(String(describing: error))"
+            )
+
+            return []
+        }
+    }
+
+    private func processMLSMessageDecryptionResults(
+        _ results: [MLSDecryptResult],
+        mlsConversation: ZMConversation,
+        senderID: UUID,
+        senderDomain: String,
+        date: Date?
+    ) async -> [ConversationMLSMessageAddEvent.DecryptedMessage] {
+        var decryptedMessages: [ConversationMLSMessageAddEvent.DecryptedMessage] = []
+
+        for result in results {
+            switch result {
+            case .message(let decryptedData, let senderClientID):
+                let mlsDecryptedMessage = decryptedData.base64EncodedString()
+                decryptedMessages.append(
+                    .init(
+                        message: mlsDecryptedMessage,
+                        senderClientID: senderClientID
+                    )
+                )
+
+            case .proposal(let commitDelay):
+                let scheduledDate = (date ?? Date.now) + TimeInterval(commitDelay)
+
+                await conversationLocalStore.storeConversation(
+                    commitPendingProposalDate: scheduledDate,
+                    conversation: mlsConversation
+                )
+
+                mlsService.commitPendingProposalsIfNeeded()
```

Comment:
done in https://github.com/wireapp/wire-ios/pull/2164/commits/c08ed17ebb783810d1b65e98d08654d10a6aaf59, revisited the current implementation

Reply:
Although I see your point, I'm not sure that makes sense in that specific case since we're iterating over the decryption results which is an enum with two tightly coupled cases `message` (the decrypted message) and `proposal` (the code here).  We need to decrypt the MLS message here but since it is an enum (I think) it makes more sense to deal with the other case at the same time. 
What I did though is grouping this logic into a single method from the local store since we already use the mls service in that store. https://github.com/wireapp/wire-ios/pull/2164/commits/c08ed17ebb783810d1b65e98d08654d10a6aaf59#diff-508a6af2dc08e794560553f4f59fd1d749c0a36db25bd806d3d25fcd7a619b70